### ISSUE TITLE
RpcError: invalid base64 string.

### ISSUE THREAD

**Issue by DiogoMaMartins** [timostamm/protobuf-ts#53]
it communicates with the server and it works but I get this error,
how can i solve i'm using react with typescript ?

 invalid base64 string.
    RpcError rpc-error.js:6
    unary grpc-web-transport.js:163
    promise callback*unary grpc-web-transport.js:160
    tail rpc-interceptor.js:10
    stackIntercept rpc-interceptor.js:15
    welcome contact.ts:516
    callGrpcWelcomeService index.tsx:31
    onClick index.tsx:72
    React 14
    unstable_runWithPriority scheduler.development.js:646
    React 15
    tsx index.tsx:7
    tsx main.chunk.js:8393
    Webpack 7

**Comment by timostamm** [comment]
grpc-web has two modes, binary and text. Both modes serialize proto in binary format. The text mode wraps the binary data in base64 encoding. 

It looks like your server sends base64 data that cannot be decoded. This is either problem on your server - or a bug in protobuf-ts.

Use the browser developer tools to see what the server is actually sending. The response content type should be "application/grpc-web-text". 

What are you using on the server side?

**Comment by DiogoMaMartins** [comment]
> grpc-web has two modes, binary and text. Both modes serialize proto in binary format. The text mode wraps the binary data in base64 encoding.
> 
> It looks like your server sends base64 data that cannot be decoded. This is either problem on your server - or a bug in protobuf-ts.
> 
> Use the browser developer tools to see what the server is actually sending. The response content type should be "application/grpc-web-text".
> 
> What are you using on the server side?

Im using golang and im returning return &pb.WelcomeRes{Status: pb.Status_Success}, nil  
my status is enum Status {
  Success = 0;
  Fail = 1;
}

**Comment by timostamm** [comment]
Sure, but your are probably using some proxy for grpc-web as well? 

Can't help you without some more info to reproduce...

**Comment by DiogoMaMartins** [comment]
> Sure, but your are probably using some proxy for grpc-web as well?
> 
> Can't help you without some more info to reproduce...

yes im using envoy and my server works well with my api i just have a err with grpc web this is my front
```
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
let transport = new GrpcWebFetchTransport({
    baseUrl: 'http://localhost:9090',
});

let client = new ContactServiceClient(transport);


const ForgotPasswordForm: React.FC = () => {
    const callGrpcWelcomeService = async () => {
      try{
        let message: EmailWelcome = await { clientName: "test",
        to:"kevip16466@febeks.com" }


        const { response } = await client.welcome({email:message});
        await console.log("response",response)
      }catch(err){
        console.log("im err",err)
      }
      } 
```
and this is my envoy.yaml

```
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address: { address: 0.0.0.0, port_value: 9901 }

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address: { address: 0.0.0.0, port_value: 9090 }
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route:
                  cluster: contact_service
                  max_grpc_timeout: 0s
              cors:
                allow_origin_string_match:
                - prefix: "*"
                allow_methods: GET, PUT, DELETE, POST, OPTIONS
                allow_headers: keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,grpc-timeout
                max_age: "1728000"
                expose_headers: custom-header-1,grpc-status,grpc-message
          http_filters:
          - name: envoy.grpc_web
          - name: envoy.cors
          - name: envoy.router
  clusters:
  - name: contact_service
    connect_timeout: 0.25s
    type: logical_dns
    http2_protocol_options: {}
    lb_policy: round_robin
    hosts: [{ socket_address: { address: host.docker.internal, port_value: 8080  }}]

```

and my server return a objecct like  response, email:{client_name:"test" to:"kevip16466@febeks.com"}

i

**Comment by timostamm** [comment]
> email:{client_name:"test" to:"[kevip16466@febeks.com](mailto:kevip16466@febeks.com)"}

This is not a valid response body in any format. Use the developer tools in your browser to see what is actually happening. 

The request should look like this:

<img width="1148" alt="req" src="https://user-images.githubusercontent.com/4289451/102712180-565c2c00-42bf-11eb-8ffb-2550c3565060.png">

Note:
- request content-type: application/**grpc-web-text**
- response content-type: application/**grpc-web-text**
- base64 encoded request payload

The response body:

<img width="1148" alt="res" src="https://user-images.githubusercontent.com/4289451/102712233-c4085800-42bf-11eb-95e9-57e0b965f6d9.png">

Note:
- base64 encoded response body


I have a hunch that envoy is not sending grpc-web-text. Please use dev tools and paste screenshots here.

**Comment by DiogoMaMartins** [comment]
> > email:{client_name:"test" to:"[kevip16466@febeks.com](mailto:kevip16466@febeks.com)"}
> 
> This is not a valid response body in any format. Use the developer tools in your browser to see what is actually happening.
> 
> The request should look like this:
> 
> <img alt="req" width="1148" src="https://user-images.githubusercontent.com/4289451/102712180-565c2c00-42bf-11eb-8ffb-2550c3565060.png">
> 
> Note:
> 
>     * request content-type: application/**grpc-web-text**
> 
>     * response content-type: application/**grpc-web-text**
> 
>     * base64 encoded request payload
> 
> 
> The response body:
> 
> <img alt="res" width="1148" src="https://user-images.githubusercontent.com/4289451/102712233-c4085800-42bf-11eb-95e9-57e0b965f6d9.png">
> 
> Note:
> 
>     * base64 encoded response body
> 
> 
> I have a hunch that envoy is not sending grpc-web-text. Please use dev tools and paste screenshots here.

you're right is returning, application/grpc-web+proto,but I don't know why

![Screenshot 2020-12-20 at 14 00 23](https://user-images.githubusercontent.com/43170665/102713928-d9837f00-42cb-11eb-991b-7a314593f75d.png)

**Comment by timostamm** [comment]
> you're right is returning, application/grpc-web+proto,but I don't know why

That's how envoy interprets the spec. I'll have a look into supporting this.

This workaround should fix your problem:

```typescript
let transport = new GrpcWebFetchTransport({
    baseUrl: 'http://localhost:9090',
    format: 'binary'
});
```

**Comment by DiogoMaMartins** [comment]
> > you're right is returning, application/grpc-web+proto,but I don't know why
> 
> That's how envoy interprets the spec. I'll have a look into supporting this.
> 
> This workaround should fix your problem:
> 
> ```ts
> let transport = new GrpcWebFetchTransport({
>     baseUrl: 'http://localhost:9090',
>     format: 'binary'
> });
> ```

Thank you very much it worked!You rock!

**Comment by timostamm** [comment]
👍 

Just released version 2.0.0-alpha.7 that includes a fix for envoy compatibility. 

You can install this pre-release with the "next" channel:

```
npm i @protobuf-ts/grpcweb-transport@next @protobuf-ts/runtime@next @protobuf-ts/runtime-rpc@next @protobuf-ts/plugin@next 
```

**Comment by bartenra** [comment]
We ran into this as well (we are also using Envoy). Nice to see it fixed in an upcoming release.
